### PR TITLE
Adding unit test for Metric component

### PR DESCRIPTION
**Pull request by namnguyen20999**
**Previous state:**

| File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s |
|------------|---------|----------|---------|---------|-------------------|
| Metric.tsx | 83.05   | 49.05    | 72.72   | 83.92   | 75-82,161,164,168 |

**New state:**

| File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s |
|------------|---------|----------|---------|---------|-------------------|
| Metric.tsx | 100     | 90.56    | 100     | 100     | 82,140,155,159-161 |

### CODE REVIEW

File: frontend/taipy-gui/src/components/Taipy/DateSelector.tsx [hunk 1]
```diff
@@ -80,6 +80,16 @@ const DateSelector = (props: DateSelectorProps) => {
         [updateVarName, dispatch, withTime, propagate, tz, props.onChange, module],
     );
 
+    const handleConsoleLog = () => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/DateSelector.tsx]: remove this ?

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx [hunk 2]
```diff
@@ -11,21 +11,415 @@
...
+    it("displays the right info for class", async () => {
+        const {getByTestId} = render(<Metric testId="test-id" className={'taipy-gauge'}/>);
+        const elt = getByTestId('test-id');
+        expect(elt).toHaveClass('taipy-gauge');
+    })
+
+    it("sets the title when provided", async () => {
+        const title = "Test Title";
+        let container: RenderResult | null = null;
+        container = render(<Metric title={title}/>);
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx]: why not use const {container}=render(...) ?

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx [hunk 3]
```diff
@@ -11,21 +11,415 @@
...
+        render(<Metric colorMap={malformedJson}/>);
+        expect(consoleSpy).toHaveBeenCalledWith(expect.stringContaining('Error parsing color_map value (metric)'));
+        consoleSpy.mockRestore();
+    });
+
+    it("sets the template when provided", async () => {
+        let container: RenderResult | null = null;
+        container = render(<Metric template={JSON.stringify(template)}/>);
+        await waitFor(() => {
+            const elmWithTpl = container!.container.querySelector('.user-select-none')
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx]: How does this test the use of the template ?

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx [hunk 4]
```diff
@@ -11,21 +11,415 @@
...
+            expect(elt).toHaveStyle({
+                backgroundColor: 'rgb(255, 255, 255)'
+            });
+        });
+    });
+
+    it('logs an error when template_Dark_ prop is not a valid JSON string', () => {
+        const consoleSpy = jest.spyOn(console, 'info');
+        render(<Metric template_Dark_="not a valid JSON string" testId="test-id"/>);
+        expect(consoleSpy).toHaveBeenCalledWith(expect.stringContaining('Error while parsing Metric.template'));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx]: I didn't know you could do this !

**Review by FredLL-Avaiga — DISMISSED** [review]

File: frontend/taipy-gui/src/components/Taipy/DateSelector.tsx (hunk 1 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/DateSelector.tsx]: I've removed it. Added for testing but I forgot to remove this file from the PR

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx (hunk 2 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx]: Updated

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx (hunk 3 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx]: I've refactored this test case to check the use of template

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx (hunk 4 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx]: This case isn't working right now, so I'll skip it. I will come back to this later. It works with template_Light_ but not with template_Dark_; I'm not sure why yet.

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx [hunk 5]
```diff
@@ -12,20 +12,321 @@
...
 import React from "react";
-import {render} from "@testing-library/react";
+import { render, waitFor } from "@testing-library/react";
 import "@testing-library/jest-dom";
+import { ThemeProvider } from "@mui/material/styles";
+import { createTheme } from "@mui/material/styles";
 
 import Metric from "./Metric";
 
+const template = {
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx:22]: Honestly, for testing purposes, I would have made those far lighter.
But well, go for it.

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx [hunk 6]
```diff
@@ -12,20 +12,321 @@
...
+    });
+
+    it.skip("logs an error when template_Dark_ prop is not a valid JSON string", () => {
+        const consoleSpy = jest.spyOn(console, "info");
+        render(<Metric template_Dark_="not a valid JSON string" testId="test-id" />);
+        expect(consoleSpy).toHaveBeenCalledWith(expect.stringContaining("Error while parsing Metric.template"));
+        consoleSpy.mockRestore();
+    }); // TODO: Not working at the moment, need to fix
+
+    it("logs an error when template_Light_ prop is not a valid JSON string", () => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx:326]: is this one working ?

**Review by FredLL-Avaiga — APPROVED** [review]

**Review by FabienLelaquais — APPROVED** [review]
I liiiike that!

File: frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx (hunk 6 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.spec.tsx:326]: The Light is working but Dark one it not, I don't know why yet so I skip it at the moment

**Pull request merged**